core-processor/src/ext.rs

```diff
@@ -319,6 +319,134 @@ impl GrowHandler for LazyGrowHandler {
     }
 }
 
+enum Change {
+    MarkUsed(ReservationId),
+}
+
+/// Used to atomically update `Ext` which prevents some errors
+/// when data was updated but operation failed.
+///
+/// Copies some counters into itself and performs operations on them and
+/// incrementally builds list of changes.
+///
+/// Changes are applied after operation is completed
+struct ExtChanges {
+    gas_counter: GasCounter,
+    gas_allowance_counter: GasAllowanceCounter,
+    value_counter: ValueCounter,
+    outgoing_glassless: u64,
+    mailbox_threshold: u64,
+    // TODO: Maybe replace with `SmallVec` to reduce heap usage?
```

Comment:
Nice idea, I'd apply it right here
```suggestion
```

Reply:
overall at this point it seems that changes could contain only one change so I'd remove it until needed